### DISCUSSION TITLE
keycloak (Quarkus) cluster

### DISCUSSION THREAD

**Discussion by AlexeiKlimenko** [keycloak/keycloak#11874] [Q&A]
Good day.
We're looking for any suitable info about implementation keycloak (quarkus) cluster on-prem (not docker). Could anyone help with it?
Should we deploy Infinispan ?
Have anyone experienced with cluster configuration (quarkus) and realized how to properly build it?

**Comment by andreaTP** [comment]
Hi @AlexeiKlimenko , thanks for reaching out!
I think your question boils down to "how to configure the embedded Infinispan".

As a starting point, you can look at the documentation: https://www.keycloak.org/server/caching
and we implemented it in the operator (for Kubernetes) here: #10387

Let me know if you have any more specific questions!

**Comment by Wolfsrudel** [comment]
I think this should be the right one:
https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80

On-prem, no container, no operator.

**Comment by AlexeiKlimenko** [comment]
Hello, guys!
Thanks for articles, but it's not still clear about using Infinispan. Is it really neccesary to be used with keycloak cluster?
Has anyone got any suitable article (at least own notes) to configure instances quarkus keycloaks as cluster?

**Reply by andreaTP** [reply]
> Is it really neccesary to be used with keycloak cluster?

Yes, clustering is based on Infinispan

> Has anyone got any suitable article (at least own notes) to configure instances quarkus keycloaks as cluster?

Yes, the configuration of Infinispan: https://www.keycloak.org/server/caching

**Comment by AlexeiKlimenko** [comment]
Thank you, Andrea.
What about additional keys should be added to start script (sudo bin/kc.sh start) if we're working with cluster instances?

**Reply by andreaTP** [reply]
it really depends on "where" you are deploying and your requirements.

Please refer to the "Relevant options" in the [Caching guide](https://www.keycloak.org/server/caching).
Please note that those are all "build time" options and they should be provided to the `kc.sh build` command.

**Comment by AlexeiKlimenko** [comment]
As mentioned above we're installing instance like bare metal (neither kuber nor docker).
We've read docs about caching but questions about additional keys are still remaining.
Actually, it seems to me, that no other keys except cache are not needed for starting, right?

**Comment by ashishtchaudhari** [comment]
@AlexeiKlimenko  -- Were you able to get this configured ? We are also trying to achieve the same thing. Few things that are not clear to me are, 
How does the clustering for Keycloak (not cache) work? Like in Wildfly we had domain mode, so that we configure only at one central place? How do we do that with Quarkus ?
